test/11_dependency_versions/cibuildwheel_test.py

```diff
@@ -0,0 +1,105 @@
+import os
+import re
+import pytest
+import textwrap
+import cibuildwheel.util
+
+import utils
+
+
+def get_version_from_constraint_file(package_name, constraint_file):
+    version_pattern = package_name + r'==([^\s]+)'
+
+    with open(constraint_file, encoding='utf8') as f:
+        constraint_file_text = f.read()
+
+    match = re.search(version_pattern, constraint_file_text)
+
+    if not match:
+        raise Exception("couldn't find version spec matching " + version_pattern)
+
+    return match.group(1)
+
+
+@pytest.mark.parametrize('python_version', ['2.7', '3.x'])
+def test_pinned_versions(python_version):
+    if utils.platform == 'linux':
+        pytest.skip('linux doesn\'t pin individual tool versions, it pins manylinux images instead')
+
+    project_dir = os.path.dirname(__file__)
+
+    build_environment = {}
+
+    if python_version == '2.7':
+        constraint_filename = 'constraints-python27.txt'
+    else:
+        constraint_filename = 'constraints.txt'
+
+    constraint_file = os.path.join(cibuildwheel.util.resources_dir, constraint_filename)
+
+    for package in ['pip', 'setuptools', 'wheel', 'virtualenv']:
+        env_name = 'EXPECTED_{}_VERSION'.format(package.upper())
+        build_environment[env_name] = get_version_from_constraint_file(package, constraint_file)
+
+    cibw_environment_option = ' '.join(['{}={}'.format(k, v) for k, v in build_environment.items()])
```

Comment:
Why first define `build_environment` and use it only to generate string? 

```python
env_opt = " ".join([ 
    'EXPECTED_{}_VERSION={}'.format(
        package.upper(), get_version_from_constraint_file(package, constraint_file))
    for package in ['pip', 'setuptools', 'wheel', 'virtualenv']
])
```

Reply:
I like this as well. If not, maybe move `build_environment = {}` down to just before the loop? Or maybe fill it with a dict-comprehension.

Reply:
I see what you mean, but personally, I prefer to build the information in a native data structure, then encode it as a second step. It feels natural to me to work with environment dictionaries.